### PR TITLE
Fixed #13158 (IntelliSense does not display get-only abstract collection properties in object initializers).

### LINKED ISSUE #13158: IntelliSense does not display get-only abstract collection properties in object initializers
**Version Used**: VS 2015 Update 3 (14.0.25425.01)

**Steps to Reproduce**:

1 - Write the following class:  

``` csharp
class C
{
    public IList<int> Items { get; }
}
```

2 - Instantiate it with an object initializer:  

``` csharp
var c = new C
{
};
```

3 - Move your cursor inside the initializer, and toggle IntelliSense (e.g. Ctrl+Space).

**Expected Behavior**: IntelliSense should display the `Items` property, since items can be added to it even though the property itself is read-only, e.g. `Items =  { 1, 2, 3 }`.

**Actual Behavior**: Nothing is displayed.

**Notes**:
- IntelliSense works as expected if `Items` is a `List<int>`, or if it has a `set` accessor.
- A real-life example of this is `Panel.Children` in XAML apps.

### PR DESCRIPTION
**Pull request by zaytsev-victor**
**Customer scenario**
IntelliSense does not display get-only abstract collection properties in object initializers

**Bugs this fixes:**
#13158

**Risk**
Low

**Performance impact**

**Is this a regression from a previous update?**
No

**Root cause analysis:**

Missed check for interface members for abstract collection properties 

**How was the bug found?**

customer reported

### CODE REVIEW

File: src/EditorFeatures/CSharpTest/Completion/CompletionProviders/ObjectInitializerCompletionProviderTests.cs [hunk 1]
```diff
@@ -701,6 +701,58 @@ static void Main(string[] args)
...
+    }
+}";
+
+            await VerifyItemExistsAsync(markup, "Items");
+            await VerifyItemExistsAsync(markup, "Bar");
+        }
+
+        [WorkItem(13158, "https://github.com/dotnet/roslyn/issues/13158")]
+        [Fact, Trait(Traits.Feature, Traits.Features.Completion)]
+        public async Task CollectionInitializerForInterfaceType2()
```
Comment (Pilchie) [inline src/EditorFeatures/CSharpTest/Completion/CompletionProviders/ObjectInitializerCompletionProviderTests.cs]: The original bug report mentions `Panel.Children` in XAML apps, but that is a `UIElementCollection` which isn't an interface.  Does this fix address that?  If so, how?  Regardless we should probably have a test for a similar scenario as that one.

File: src/Workspaces/Core/Portable/Shared/Extensions/ITypeSymbolExtensions.cs [hunk 2]
```diff
@@ -722,6 +723,16 @@ public static INamedTypeSymbol GetDelegateType(this ITypeSymbol typeSymbol, Comp
...
         }
 
+        private static IEnumerable<T> SelectAccessibleMembers<T>(this IEnumerable<ITypeSymbol> types, ISymbol within) where T : class, ISymbol
+        {
+            if (types == null)
+            {
+                return ImmutableArray<T>.Empty;
+            }
+
+            return types.SelectMany(x => x.GetMembers().OfType<T>().Where(m => m.IsAccessibleWithin(within)));
```
Comment (Pilchie) [inline src/Workspaces/Core/Portable/Shared/Extensions/ITypeSymbolExtensions.cs]: Slight nit that existed in the original to, but we could pass `name` to `GetMembers` to make this not need to create and enumerate as many symbols.

File: src/EditorFeatures/CSharpTest/Completion/CompletionProviders/ObjectInitializerCompletionProviderTests.cs (hunk 1 above)
Comment (zaytsev-victor) [inline src/EditorFeatures/CSharpTest/Completion/CompletionProviders/ObjectInitializerCompletionProviderTests.cs]: I don't know how to reproduce bug in XAML app.
This behavior is observer in VS2017 15.2
xaml_panel_children

File: src/Workspaces/Core/Portable/Shared/Extensions/ITypeSymbolExtensions.cs (hunk 2 above)
Comment (zaytsev-victor) [inline src/Workspaces/Core/Portable/Shared/Extensions/ITypeSymbolExtensions.cs]: Added overload for ```SelectAccessibleMembers``` with ```name``` parameter.

**Comment by Pilchie** [comment]
retest windows_debug_vs-integration_prtest please

**Comment by Pilchie** [comment]
Results were gone :(

File: src/EditorFeatures/CSharpTest/Completion/CompletionProviders/ObjectInitializerCompletionProviderTests.cs (hunk 1 above)
Comment (Pilchie) [inline src/EditorFeatures/CSharpTest/Completion/CompletionProviders/ObjectInitializerCompletionProviderTests.cs]: I suspect that the issue is related to UWP, and Windows Runtime Projections.  I verified that it *does* repro without your fix in 15.3 for a UWP app, but *does not* repro with your fix, so I'm happy with this.

**Comment by Pilchie** [comment]
@dotnet/roslyn-infrastructure https://ci.dot.net/job/dotnet_roslyn/job/master/job/windows_debug_vs-integration_prtest/6405/ passed, but failed to notify helix.  Merging with Admin privs anyway.

**Comment by Pilchie** [comment]
Thanks @zaytsev-victor !

**Pull request merged**